lib/node_modules/@stdlib/math/base/special/boxcox/test/test.js

```diff
@@ -0,0 +1,110 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var tape = require( 'tape' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var randu = require( '@stdlib/random/base/randu' );
+var abs = require( '@stdlib/math/base/special/abs' );
+var NINF = require( '@stdlib/constants/math/float64-ninf' );
+var EPS = require( '@stdlib/constants/math/float64-eps' );
+var boxcox = require( '@stdlib/math/base/special/boxcox' );
+
+
+// FIXTURES //
+
+var data = require( './fixtures/python/data.json' );
+
+
+// TESTS //
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.equal( typeof boxcox, 'function', 'main export is a function');
+	t.end();
+});
+
+tape( 'the function returns `NaN if provided `NaN`', function test( t ) {
+	var b;
+
+	b = boxcox( NaN, 1);
+	t.equal( isnan( b ), true, 'returns NaN' );
+	b = boxcox( 1, NaN );
+	t.equal( isnan( b ), true, 'returns NaN' );
+	t.end();
+});
+
+tape( 'the function returns `-Infinity` if lambda is negative', function test( t ) {
+	var iterations;
+	var b;
+	var x;
+	var i;
+
+	iterations = 1000;
+	for ( i = 0; i < iterations; i++ ) {
+		x = -1 * (( randu() * 100.0 ) + 1);
+		b = boxcox( 2.3, x );
+		t.strictEqual( b, NINF, 'returns -Infinity when lambda is negative.');
+	}
+	t.end();
+});
+
+tape( 'the function returns `NaN` if number (x) is negative', function test( t ) {
+	var iterations;
+	var b;
+	var x;
+	var y;
+	var i;
+
+	iterations = 1000;
+	for ( i = 0; i < iterations; i++ ) {
+		x = -1 * (( randu() * 100.0 ) + 1);
+		y = ( randu() * 10.0 );
+		b = boxcox( x, y );
+		t.strictEqual( b, NaN, 'returns NaN');
+	}
+	t.end();
+});
+
+tape( 'the function accurately computes the Box-Cox transformation.', function test( t ) {
+	var delta;
+	var tol;
+	var ex;
+	var i;
+	var b;
+	var x;
+	var y;
+
+	for ( i = 0; i < data.length; i++ ) {
+		x = data[i].input[0];
+		y = data[i].input[1];
+		ex = data[i].expected;
+		b = boxcox( x, y );
+		if ( b === ex ) {
+			t.strictEqual( b, ex, 'returns '+b+'when provided '+x+' and '+y+'.');
+		} else {
+			delta = abs( ex - b );
+			tol = EPS * abs( ex );
```

Comment:
In fixing the testing issues, I have a couple of tests that are failing because of tolerance issues. If I change to `@stdlib/constants/math/float32-eps` instead of `@stdlib/constants/math/float64-eps`, then the tests pass. What is the recommended course of action regarding these floating point comparison?

Reply:
Sorry for the confusion here. We really need to document this, as it has been a recurring question. The rationale for L105 is that we want to get a handle on how much different our implementations are from reference implementations in terms of ULPs (or the number of trailing bits which differ in the significand). Our proxy to this (for now) is to scale double-precision floating-point epsilon. Currently, the L105 suggests that the tolerance is approximately one ULP (or within floating-point rounding error). When tests fail, we scale `EPS` by some real-valued multiple (e.g., `2.0 * EPS * abs( ex )`) until tests pass. This does not have to be exact, but you can get a rough idea of how large the tolerance needs to be by looking at the test message output which should include both `delta` and `tol`. Based on those values, we can derive an approximate "minimum" tolerance required for tests to pass.

Obviously, this is a bit ad hoc, and we should invest some time into automating this.